rfcs/0079-include-key-as.md

```diff
@@ -0,0 +1,89 @@
+---
+feature: New Syntax Include Key As
+authors:
+  - "TabVersion"
+start_date: "2023/11/22"
+---
+
+# New Syntax: Include \<column\> as \<new name\>
+
+## Background
+
+> This RFC is a follow-up to [RFC-0064](./0064-new-source-ddl.md).
+
+In previous implementation, we only load message key into a [~~hidden~~](https://github.com/risingwavelabs/risingwave/pull/13521) 
+column called `_rw_key` as `bytea` on default.
+
+Here are some cases when RisingWave derives the `_rw_key` column.
+
+* For all connectors with `format upsert`, RisingWave derives the column as primary key to perform upsert semantic.
+  * An exception: for `format upsert encode json`, RisingWave allows the PK to be part of the message payload instead of
+  the message key. This behavior is more like `format plain encode json` with PK constraint.
+  * **To Be Discussed**: whether we can afford to perform a breaking change for the above exception.
+* For all connectors with `format plain`, RisingWave derives the `_rw_key` column as a normal one
+  ([ref](https://github.com/risingwavelabs/risingwave/pull/13278)) at a POC request. This enforces an additional column
+  for all tables, which is not ideal.
+
+Aside from message keys, users may want to ingest the timestamp as well as the topic headers into the table for further
+analysis.
+
+This RFC proposes a new syntax to allow users to specify the column name for message components other than payload.
+
+## Proposed Syntax
+
+```sql
+create source/table (
+    ..,
+    primary key ( <key-column> )
+)
+    [ include key [ as <key-column> ] ]
+    [ include timestamp [ as <timestamp-column>] ]
+    [ include header [as <header-column>] ]
+    [ include ... [ as ... ] ]
```

Comment:
Have you considered the generated column syntax for this? For example, calling dummy functions to get the meta data.

Reply:
Generated column asks users to specify the data type, which might be a little bit challenging sometimes, such as https://github.com/risingwavelabs/rfcs/pull/79#discussion_r1402891627